### PR TITLE
[wip] Remove multissid and enable multi instance

### LINKED ISSUE #100: Remove multi SSID feature
During the YunoCamp with @alexAubin, we were wondering if we should keep the multi SSID feature (i.e. having more than one Wifi hotspot).
This feature makes the code really complex for what I guess is a pretty rare use case. In general, people have a single SSID and that's it…

What are your thoughts about this? Should we keep this feature?

### PR DESCRIPTION
**Pull request by hidrarga**
## Problem

Closes #100

## Solution

While I'm removing the multi SSID to make the app cleaner and simpler, I'm also enabling installation of multiple instances of the app. The user will be able to install one hotspot app per wifi device. The downside is that they won't be able to have multiple SSID on the same device (we could add this feature later, for instance by adding a "guest hotspot" option).

## PR Status

- [ ] Code finished and ready to be reviewed/tested
- [ ] The fix/enhancement were manually tested (if applicable)

## Automatic tests

Automatic tests can be triggered on https://ci-apps-dev.yunohost.org/ *after creating the PR*, by commenting "!testme", "!gogogadgetoci" or "By the power of systemd, I invoke The Great App CI to test this Pull Request!". (N.B. : for this to work you need to be a member of the Yunohost-Apps organization)

### CODE REVIEW

**Comment by hidrarga** [comment]
!testme

**Comment by hidrarga** [comment]
!testme

**Comment by hidrarga** [comment]
!gogogadgetoci

File: scripts/upgrade [hunk 1]
```diff
@@ -68,10 +80,68 @@ elif [ $firmware_nonfree = "no" ]; then
 fi
 
 if [ -z $service_name ]; then
-	service_name="ynh-hotspot"
+	service_name="ynh-$app"
 	ynh_app_setting_set --app=$app --key=service_name --value=$service_name
 fi
 
+if [[ -n ${multissid} ]] && [[ ${multissid} -gt 1 ]]; then
```
Comment (alexAubin) [inline scripts/upgrade:87]:
```suggestion
if [[ -n "${multissid}" ]] && [[ "${multissid}" -gt 1 ]]; then
```

File: scripts/upgrade [hunk 2]
```diff
@@ -68,10 +80,68 @@ elif [ $firmware_nonfree = "no" ]; then
...
+		ynh_app_setting_set --app=$app --key=ip4_dns --value="${ip4_dns}"
+	fi
+
+	ynh_app_setting_delete $app dns
+else
+    ip6_dns=$(ynh_app_setting_get --app=$app --key=ip6_dns)
+    ip4_dns=$(ynh_app_setting_get --app=$app --key=ip4_dns)
+fi
+
+if [[ -n ${multissid} ]]; then
```
Comment (alexAubin) [inline scripts/upgrade:132]:
```suggestion
if [[ -n "${multissid}" ]]; then
```

File: scripts/upgrade [hunk 3]
```diff
@@ -68,10 +80,68 @@ elif [ $firmware_nonfree = "no" ]; then
...
+    ynh_app_setting_delete --app=$app --key=multissid
+
+	ynh_secure_remove --file="/etc/hostapd/hostapd.conf"
+	ynh_secure_remove --file="/etc/hostapd/hostapd.base.conf"
+	ynh_secure_remove --file="/etc/hostapd/hostapd.accesspoint.conf"
+	ynh_secure_remove --file="/etc/dnsmasq.dhcpd/"
+fi
+
+if [[ -z ${advanced} ]]; then
```
Comment (alexAubin) [inline scripts/upgrade:141]:
```suggestion
if [[ -z "${advanced}" ]]; then
```

File: scripts/upgrade [hunk 4]
```diff
@@ -131,15 +201,39 @@ ynh_install_app_dependencies $pkg_dependencies
 #=================================================
 ynh_script_progression --message="Copying configuration..."
 
-mkdir -pm 0755 /etc/dnsmasq.dhcpd/
-chown root: /etc/dnsmasq.dhcpd/
+hot_reload_usb_wifi_cards
+if [[ -z $wifi_device ]] || ! grep -q -F $wifi_device <(unused_iw_devices); then
```
Comment (alexAubin) [inline scripts/upgrade:205]:
```suggestion
if [[ -z "$wifi_device" ]] || ! grep -q -F "$wifi_device" <(unused_iw_devices); then
```

File: scripts/upgrade [hunk 5]
```diff
@@ -131,15 +201,39 @@ ynh_install_app_dependencies $pkg_dependencies
...
+	wifi_device=$(unused_iw_devices | head -n 1)
```
Comment (alexAubin) [inline scripts/upgrade:206]:
```suggestion
	wifi_device="$(unused_iw_devices | head -n 1)"
```

File: scripts/upgrade [hunk 6]
```diff
@@ -131,15 +201,39 @@ ynh_install_app_dependencies $pkg_dependencies
...
-install -b -o root -g root -m 0644 ../conf/dnsmasq_dhcpdv6.conf.tpl /etc/dnsmasq.dhcpd/dhcpdv6.conf.tpl
-install -b -o root -g root -m 0644 ../conf/dnsmasq_dhcpdv4.conf.tpl /etc/dnsmasq.dhcpd/dhcpdv4.conf.tpl
+mkdir -pm 0755 /etc/dnsmasq.$app/
+chown root: /etc/dnsmasq.$app/
+
+install -b -o root -g root -m 0644 ../conf/hostapd.conf.tpl /etc/hostapd/$app/hostapd.conf.tpl
+install -b -o root -g root -m 0644 ../conf/dnsmasq_dhcpdv6.conf.tpl /etc/dnsmasq.$app/dhcpdv6.conf.tpl
+install -b -o root -g root -m 0644 ../conf/dnsmasq_dhcpdv4.conf.tpl /etc/dnsmasq.$app/dhcpdv4.conf.tpl
+
+if [[ -n ${wifi_device} ]]; then
```
Comment (alexAubin) [inline scripts/upgrade:220]:
```suggestion
if [[ -n "${wifi_device}" ]]; then 
```

File: scripts/restore [hunk 7]
```diff
@@ -88,22 +92,25 @@ ynh_install_app_dependencies $pkg_dependencies
...
+yunohost service add $service_name --description "Creates a Wi-Fi access point" --test_status "systemctl is-active hostapd@$app"
 
 #=================================================
 # START SYSTEMD SERVICE
 #=================================================
 ynh_script_progression --message="Starting a systemd service..."
 
 hot_reload_usb_wifi_cards
-wifi_device=$(iw_devices | awk -F\| '{ print $1 }')
+if [[ -z $wifi_device ]] || ! grep -q -F $wifi_device <(unused_iw_devices); then
```
Comment (alexAubin) [inline scripts/restore:110]:
```suggestion
if [[ -z "$wifi_device" ]] || ! grep -q -F "$wifi_device" <(unused_iw_devices); then
```

File: scripts/install [hunk 8]
```diff
@@ -155,51 +165,62 @@ fi
...
-echo "DAEMON_CONF=/etc/hostapd/hostapd.conf" > /etc/default/hostapd
-
 # Set default inits
 # The boot order of these services are important, so they are disabled by default
 # and the ynh-hotspot service handles them.
 systemctl disable hostapd --quiet 2>&1
 systemctl stop hostapd 2>&1
 systemctl unmask hostapd 2>&1 # On some system e.g. RPi, for some reason hostapd is masked after install ...
 
+if [[ -n ${wifi_device} ]]; then
```
Comment (alexAubin) [inline scripts/install:203]:
```suggestion
if [[ -n "${wifi_device}" ]]; then 
```

File: scripts/config [hunk 9]
```diff
@@ -92,127 +93,111 @@ EOF
...
+
+    if [[ -n ${ip6_net} ]] && [[ -z ${ip6_dns} ]]; then
+        ip6_dns="${ip6_net}1"
     fi
-    local values="$(ynh_app_setting_get $app $short_setting | awk "BEGIN{FS=OFS=\"|\"} {\$${index}=\"${value}\"}"1)"
-    ynh_app_setting_set --app=$app --key=$short_setting --value="$values"
-    ynh_print_info --message="Configuration key '$short_setting' edited in app settings"
-}
 
+    if [[ -n ${ip4_nat_prefix} ]] && [[ -z ${ip4_dns} ]]; then
```
Comment (alexAubin) [inline scripts/config:171]:
```suggestion
    if [[ -n "${ip4_nat_prefix}" ]] && [[ -z "${ip4_dns}" ]]; then
```

File: scripts/config [hunk 10]
```diff
@@ -92,127 +93,111 @@ EOF
...
+            ip6_dns+="[$ip],"
+        else
+            ip4_dns+="$ip,"
+        fi
+    done
+    # Remove trailing ,
+    ip6_dns="${ip6_dns%%,}"
+    ip4_dns="${ip4_dns%%,}"
+
+    if [[ -n ${ip6_net} ]] && [[ -z ${ip6_dns} ]]; then
```
Comment (alexAubin) [inline scripts/config:167]:
```suggestion
    if [[ -n "${ip6_net}" ]] && [[ -z "${ip6_dns}" ]]; then
```

File: scripts/config [hunk 11]
```diff
@@ -92,127 +93,111 @@ EOF
...
-    local index="${1#*__}"
-    IFS='|' read -a values <<< "$(ynh_app_setting_get $app $short_setting)"
-    echo "value: \"${values[$(($index - 1))]:-}\""
+get__dns() {
+    ip6_net=$(ynh_app_setting_get --app=$app --key=ip6_net)
+    ip6_dns=$(ynh_app_setting_get --app=$app --key=ip6_dns | tr -d '[]')
+    ip4_nat_prefix=$(ynh_app_setting_get --app=$app --key=ip4_nat_prefix)
+    ip4_dns=$(ynh_app_setting_get --app=$app --key=ip4_dns)
+
+    if [[ -n ${ip6_net} ]] && [[ -z ${ip6_dns} ]]; then
```
Comment (alexAubin) [inline scripts/config:102]:
```suggestion
    if [[ -n "${ip6_net}" ]] && [[ -z "${ip6_dns}" ]]; then
```

File: scripts/config [hunk 12]
```diff
@@ -92,127 +93,111 @@ EOF
...
+        ip6_dns="${ip6_net}1"
+    fi
+
+    if [[ -n ${ip4_nat_prefix} ]] && [[ -z ${ip4_dns} ]]; then
```
Comment (alexAubin) [inline scripts/config:106]:
```suggestion
    if [[ -n "${ip4_nat_prefix}" ]] && [[ -z "${ip4_dns}" ]]; then
```

File: scripts/config [hunk 13]
```diff
@@ -76,14 +76,15 @@ EOF
 }
 
 get__wifi_device() {
-    if [[ $(iw_devices) == "" ]]
+    local unused_wifi_devices=$(unused_iw_devices)
+    if [[ -z ${unused_wifi_devices} ]]
```
Comment (alexAubin) [inline scripts/config:80]:
```suggestion
    if [[ -z "${unused_wifi_devices}" ]]
```

File: scripts/config [hunk 14]
```diff
@@ -54,15 +54,15 @@ ask:
   en: Your Hotspot is running, but it shouldn't !
 EOF
         fi
-    elif [ $service_enabled -eq 1 ]
+    elif [[ $service_enabled -eq 1 ]]
```
Comment (alexAubin) [inline scripts/config:57]:
```suggestion
    elif [[ "$service_enabled" -eq 1 ]]
```

File: scripts/config [hunk 15]
```diff
@@ -37,9 +37,9 @@ get__no_antenna() {
 
 get__status() {
     local service_enabled=$(ynh_app_setting_get $app service_enabled)
-    if systemctl is-active hostapd -q
+    if systemctl is-active hostapd@$app -q
     then
-        if [ $service_enabled -eq 1 ]
+        if [[ $service_enabled -eq 1 ]]
```
Comment (alexAubin) [inline scripts/config:42]:
```suggestion
        if [[ "$service_enabled" -eq 1 ]]
```

File: scripts/config [hunk 16]
```diff
@@ -27,7 +27,7 @@ final_path=$(ynh_app_setting_get $app final_path)
 #=================================================
 
 get__no_antenna() {
-    if [[ $(iw_devices) == "" ]]
+    if [[ $(unused_iw_devices) == "" ]]
```
Comment (alexAubin) [inline scripts/config:30]:
```suggestion
    if [[ "$(unused_iw_devices)" == "" ]]
```

File: scripts/_common.sh [hunk 17]
```diff
@@ -14,7 +14,27 @@ free_firmware_packages="firmware-ath9k-htc"
...
+
+function used_iw_devices()
+{
+  local app_shortname="${app%%__*}"
+  local hotspot_apps=$(yunohost app list --output-as json | jq -r .apps[].id | grep -F $app_shortname)
+  # Remove this app from hotspot apps list
+  local other_hotspot_apps=$(grep -F -x -v $app <<< ${hotspot_apps})
+  for hotspot_app in ${other_hotspot_apps}; do
+    hotspot_wifi_device=$(ynh_app_setting_get --app=$hotspot_app --key=wifi_device)
+    if [[ -n ${hotspot_wifi_device} ]]; then
```
Comment (alexAubin) [inline scripts/_common.sh:28]:
```suggestion
    if [[ -n "${hotspot_wifi_device}" ]]; then
```

**Review by alexAubin — COMMENTED** [review]
LGTM except i'm a bit paranoid with quoting vars in `if` statements because stuff like `if [[ -n $foo ]]` may explode when foo is empty maybe ...

**Pull request merged**